### ISSUE TITLE
This saved me countless hours

### ISSUE THREAD

**Issue by ZacharyjrWhite** [Cass-dev-web/CSHARPMSR605X#1]
There's no issue to report; I just wanted to advise that your code has saved me countless hours integrating with the MSR605x.
Also; I fixed your readCard function to read all tracks appropriately,

Thanks again.

**Comment by Cass-dev-web** [comment]
That's great to hear, I honestly was pretty suprised by how there's not a lot of code surrounding this device already. I'm happy to have helped out. Cheers!

**Comment by ZacharyjrWhite** [comment]
My original scope was to build it in python, but there are so many driver restrictions around HID integrations in python.
I'm honestly shocked that there's almost zero useful code for this device, and even less documentation.

You code made my day. It was almost plug and play, aside from the track data bug.
I also noticed you had a note regarding the writeCard function; stating that the Command hex for write didn't work. I'll let you know if I figure that out as well. 

The fix for the read card track data is below.

```c#
public ReadCardInformation ReadCard()
{
    byte[] returns = ReadCardRaw();
    if (returns.Length == 0)
    {
        ReadCardInformation errnoCard = new ReadCardInformation();
        errnoCard.failed = true;
        errnoCard.failmessage = "empty_return";
        return errnoCard;
    }
    returns = returns.Skip(3).ToArray();
    bool containsStatusCode = ContainsSubSequence(returns, new byte[] { 0x3F, 0x1C, 0x1B });
    if (!containsStatusCode)
    {
        Console.WriteLine("DOES NOT CONTAIN STATUS!!");
        ReadCardInformation errnoCard = new ReadCardInformation();
        errnoCard.failed = true;
        errnoCard.failmessage = "NO STATUS CODE ; EMPTY";
        return errnoCard;
    }
    byte statusCode = attemptFetchStatusCode(returns);
    MSRStatus status = TranslateStatusByte(statusCode);
    if (status != MSRStatus.OK)
    {
        Console.WriteLine("ERROR TRYING TO READ CARD: " + status);
        ReadCardInformation errnoCard = new ReadCardInformation();
        errnoCard.failmessage = "Hardware error: " + status;
        errnoCard.failed = true;
        return errnoCard;
    }
    ReadCardInformation readCardInformation = new ReadCardInformation();
    readCardInformation.status = status;



    int endTrack1 = -1;
    int endTrack2 = -1;
    int endTrack3 = -1;

    for (int i = 0; i < returns.Length; i++)
    {
        byte item = returns[i];
        
        uint decval = System.Convert.ToUInt32(item);
        char character = System.Convert.ToChar(decval);

        if (decval == 63 && endTrack1 == -1)
        {
            endTrack1 = i;
            i += 4;
        }
        else if (decval == 63 && endTrack2 == -1)
        {
            endTrack2 = i;
            i += 4;
        }
        else if (decval == 63 && endTrack3 == -1)
        {
            endTrack3 = i;
            break;
        }
    }

    byte[] track1ByteArray = new byte[endTrack1];
    Array.Copy(returns, 0, track1ByteArray, 0, track1ByteArray.Length);
    readCardInformation.Track1 = ConvertByteArrayToString(track1ByteArray);
    readCardInformation.Track1ByteArray = track1ByteArray;
    //
    byte[] track2ByteArray = new byte[endTrack2 - endTrack1 - 4];
    Array.Copy(returns, endTrack1 + 4, track2ByteArray, 0, track2ByteArray.Length);
    readCardInformation.Track2 = ConvertByteArrayToString(track2ByteArray);
    readCardInformation.Track2ByteArray = track2ByteArray;
      
    byte[] track3ByteArray = new byte[endTrack3 - endTrack2 - 4];
    System.Array.Copy(returns, endTrack2 + 4, track3ByteArray, 0, track3ByteArray.Length);
    readCardInformation.Track3 = ConvertByteArrayToString(track3ByteArray);
    readCardInformation.Track3ByteArray = track3ByteArray;


    return readCardInformation;
}
```

**Comment by Cass-dev-web** [comment]
> I also noticed you had a note regarding the writeCard function; stating that the Command hex for write didn't work. I'll let you know if I figure that out as well.

Thank you! I've pretty much given up because I have had already spent so much time on the project. I hope this helps you and I hope we're able to create code which can help other people too. 

Keep me posted.

**Comment by ZacharyjrWhite** [comment]
Slightly confused.
I have a demo program that came with the device. When I run the write command, I was able to use WireShark to sniff the packet that was sent.

[ 1B 6E 1B 73          1B 01 07 45 5A 69 4E FB 30 01        1B 02 04 2B 88 F9 09        1B 03 04 8B 54 FB 27 3F 1C ]

 1B 6E  - Write raw command
 1B 73  - Datablock Start
 
1B - ESC
01 - TRACK 1
07 - TRACK 1 Length
45 5A 69 4E FB 30 01 - Track 1 data

1B 02 04 
1B - ESC
02 - TRACK 2
04 - TRACK 2 Length
2B 88 F9 09 - Track 2 Data

1B 03 04
1B - ESC
03 - TRACK 3
04 - TRACK 3 Length
8B 54 FB 27 - Track 3 Data

3F 1C- Ending // Status?


My adjustment to the Write function is currently sending

[1B 6E 1B 73          1B 01 04 54 45 53 54                 1B 02 03 31 32 33           1B 03 03 34 35 36 3F 1C]

 1B 6E  - Write raw command
 1B 73  - Datablock Start

1B 01 04 - Track one, length
 54 45 53 54 - Track 1 data, etc.
 
 However, nothing happens.
 If I add D2 to the start of the byte array; the device responds with an invalid status code.
 
 Any thoughts?

**Comment by ZacharyjrWhite** [comment]
When D2 is sent as the initial byte, The response is
[ C2 1B 41 1B 01 08 A2 5A 96 72 DF 0C 80 00 1B 02 05 D4 11 9F 90 00 1B 03 05 D1 2A DF E4 00 3F 1C 1B 30 93 48 88 FE 6C B7 88 FE 6C B7 88 FE 6C B7 88 FE 6C B7 88 FE 6C B7 88 FE 6C B7 88 FE 6C B7 ]

**Comment by ZacharyjrWhite** [comment]
Key part: "1B 41                1B 01 08 A2 5A 96 72 DF 0C 80 00     1B 02 05 D4 11 9F 90 00     1B 03 05 D1 2A DF E4 00        3F 1C 1B"
1B 41 in the docs states failure; but only on some functions.
That rest seems to be track data?

**Comment by Cass-dev-web** [comment]
Yeah the [**only** documentation](https://www.triades.net/downloads/MSR605%20Programmer's%20Manual.pdf) available is pretty confusing, I would probably reverse-engineer the [program](https://msr605x.com/) which works with the MSR605X to figure out what bytes are sent to the device but that would be MAJORLY complicated, so I'd recommend doing as you said, intercepting the packets sent to the device using wireshark.

**Comment by ZacharyjrWhite** [comment]
I tried that, it's written in Delphi and I couldn't find a working decompiler online for it. This is going to be the death of me. 
It sucks there's nothing online for this. You'd think someone would have run into situations involving this.

**Comment by ZacharyjrWhite** [comment]
```c#
SendByteCommandNoReturn(new byte[] { 0xC2, 0x1B, 0x61, 0x44, 0xF8, 0x19 }); // what??? this doesn't make sense when it comes to the programmer manual :/
````

Where did you find the 0xC2 in the above, and the 0xC5 in this - I'm not seeing it in the documentation

````c#
SendByteCommandNoReturn(new byte[] { 0xC5, 0x1B, commandDictionary["RED_ONLY"] });
````

**Comment by Cass-dev-web** [comment]
Good question, I believe I might've inserted these bytes in desperation to getting it working, they might be useless but there should still be a reason why I inserted them, it might be from when I was trying to intercept the packets while using wireshark, but, once again, it might be completely useless.

**Comment by ZacharyjrWhite** [comment]
That's kind of the situation I'm in now.
Currently randomly trying things.
I wonder if the 0xC5//0xC2//etc is the report//usage location. Hmmm.
Sending the command for write doesn't seem to do anything / return anything unless 0xD2 is attached, and then it just returns the same data with the 0xC2.
I even tried dumping the entire packet that wireshark pulled into it, without any luck.

**Comment by Cass-dev-web** [comment]
Hey, I got more information. When I coded this repository my coding skills were pretty bad, so I decided to give it another try and using this very easy piece of code, I'm able to read the entire card hex for hex. I've also simplified a lot of the functions in that class. Here it is:
```c#
using HidSharp;

HidDevice[] ConnectedDevices = DeviceList.Local.GetHidDevices().ToArray();
HidDevice? MSR605X = null;
foreach (HidDevice connectedDevice in ConnectedDevices)
{
    if (connectedDevice is { ProductID: 3, VendorID: 2049 })
        MSR605X = connectedDevice;
}

if (MSR605X == null)
    throw new Exception("MSR not found!!");

HidStream MSRStream = MSR605X.Open();
MSRStream.ReadTimeout=Int32.MaxValue;
MSRStream.WriteTimeout=Int32.MaxValue;
MSRStream.Closed += (_,_) => Console.WriteLine($"** stream closed **");

byte[] CreateReportData(List<byte> data, int length = 64)
{
    // Create a report to send from a hex dump.
    if (data.Count < length)
    {
        List<byte> remaining = new List<byte>(new byte[length - data.Count]);
        data = new List<byte>(data);
        data.AddRange(remaining);
    }

    data=data.Prepend((byte)0xC2).ToList(); // Prepend the secondary interface code 0xC2
    data=data.Prepend((byte)0x00).ToList(); // Prepend the interface code 0x00
    // DON'T ASK ME WHY, BUT THIS WORKS! Also, no clue why or WHERE this came from.
    return data.ToArray();
}

MSRStream.SetFeature(CreateReportData(new List<byte>() { 0x1B, 0x61 }));
Thread.Sleep(1000);
MSRStream.SetFeature(CreateReportData(new List<byte>() { 0x1B, 0x72 }));
while (true)
{
    byte[] buff = MSRStream.Read();
    if (buff.Length == 0)
        continue;
    foreach (var VARIABLE in buff)
    {
        Console.Write(VARIABLE.ToString("X2") + " ");
    }
    Console.WriteLine();
    MSRStream.Flush();
    break;
}
MSRStream.Close();
```

**Comment by Cass-dev-web** [comment]
Using a bit more UTF-8 formatting I'm able to find the data associated to my card:
<img width="1061" alt="image" src="https://github.com/Cass-dev-web/CSHARPMSR605X/assets/60286224/fbc3b636-9abb-4718-9e12-11b5e3e96012">
**...X X X...** and **...9 9 9...**
```c#
foreach (var VARIABLE in buff)
    {
        if(VARIABLE==0x1B)
            Console.Write(VARIABLE.ToString("X2") + " ");
        else
            Console.Write(System.Text.Encoding.UTF8.GetString(new byte[]{VARIABLE})+" ");
    }
```

**Comment by ZacharyjrWhite** [comment]
Hey that’s exciting information, sorry I’m not at my computer to run this and reading code on mobile is distracting. Were you able to write to cards?

**Comment by Cass-dev-web** [comment]
Yeah I haven't tried writing yet since it would require me to write a small parser for the text, I'll get on it as soon as i complete the read parser.

**Comment by ZacharyjrWhite** [comment]
That’s fair! let me know what you find. 
I’ve sidelined this project as its the final piece of a larger project I’m working on. 
So I wasn’t expecting to hear more progress being made on this for a while haha.

**Comment by Cass-dev-web** [comment]
Sounds good! Don't worry I'll try my best 💯  Currently writing a new API from scratch

**Comment by Cass-dev-web** [comment]
Okay just committed a big change introducing the second version of the interface API which can read and output cards to strings. However I have to keep testing as random strings occur and error detection is not yet implemented. Keep you updated.

update: new commit, fully fixed.
update 2: track 3 bound limit fixed as well. the card reader function is officially fully operative (except for error detection)

**Comment by Cass-dev-web** [comment]
Today has been the day. I have finally figured out how to **write data to the card** using C#.
An important aspect of this is making sure to set the coercivity before writing. 
These are the bytes sent:
```c#
0x00, 0xdd, 0x1b, 0x6e, 0x1b, 0x73,
    0x1b, 0x01, 0x09, 0x45, 0x9a, 0x2c,
    0x3c, 0x5a, 0x03, 0xa3, 0x1f, 0x32,
    0x1b, 0x02, 0x05, 0xcb, 0x66, 0x22,
    0xe0, 0x6f, 0x1b, 0x03, 0x00, 0x3f,
    0x1c 
```
This does use the raw card write method but I found that this is what the program uses when snooping around on wireshark.

I will be making a big commit soon.

**Comment by ZacharyjrWhite** [comment]
Super exciting! looks like I didn't copy enough out of wireshark when I was digging around.

**Comment by Cass-dev-web** [comment]
Alright, I pushed my commits, you may now use the code, just remember to use CSHARPMSR605X_2.cs instead of the first one. It still has some kinks to work out (for example the code not stopping for a card swipe) but I'll figure that out along the way.

**Comment by ZacharyjrWhite** [comment]
Interesting stuff! I'm curious if there's a status flag that changes throughout the 'wait for card read' // 'wait for card write' portion. Hmm

**Comment by ZacharyjrWhite** [comment]
I've been messing around with this slightly today, since it's the part of the project where I'm working on the read/write cards.
I'm wondering if you've made any more progress with the status bytes on writing.

I just ran into it returning "OK" status for writing, without actually writing to the card.
Also, swiping cards backwards returns "OK" sometimes, and "INVALID SWIPE" other times.

**Comment by Cass-dev-web** [comment]
Its honestly a bit weird, the machine seems to be kind of bipolar with what bytes it returns, I wish I could look more into this however I don't know what I haven't tried...

**Comment by ZacharyjrWhite** [comment]
There has to be something we're missing.
The application that comes with the devices handles ALL cases correctly. Example, swiping the card the wrong way.
I've got it to the point of continuous reading and writing capabilities, as well as waiting for the write status response before A: Trying again, or B: completing the write.

My write to card function is essentially
```c#
 var res = CSHARPMSR605X_2.Status.WAITING;
 while (res != CSHARPMSR605X_2.Status.OK)
 {
      res = (CSHARPMSR605X_2.Status)api.WriteCardISO(track1, track2, track3);
      Console.WriteLine(res);
 }
```

**Comment by ZacharyjrWhite** [comment]
https://github.com/Cass-dev-web/CSHARPMSR605X/assets/46058341/13564958-e1c6-4449-801c-88f7ee6c8046

Kinda got a POC situated.

**Comment by ZacharyjrWhite** [comment]
I've got it to a state where the application is now entirely asynchronous.
Reading and Writing cards now waits appropriately, until a response is gained, or a 'reset command' is sent.

I appreciate the work you've done on this immensely. I've made sure to include a reference to your github for any of the code I've used.

**Comment by ZacharyjrWhite** [comment]
Do you have discord. I'd like to have a discussion away from a public issue.
If so, add me, 
`ancestrix`

**Comment by Cass-dev-web** [comment]
All of this sounds great! And sure I'll add you on discord, check your friend requests
